R/fin_year.R

```diff
@@ -17,14 +17,20 @@
 #' @export
 fin_year <- function(date) {
 
-  if (!inherits(date, "Date")) {
-    stop("The input must have Date class.")
+  if (!inherits(date, c("Date", "POSIXct"))) {
+    stop("The input must have Date or POSIXct class.")
   }
 
-  paste0(ifelse(lubridate::month(date) >= 4,
-                lubridate::year(date),
-                lubridate::year(date) - 1), "/",
-         substr(ifelse(lubridate::month(date) >= 4,
-                       lubridate::year(date) + 1,
-                       lubridate::year(date)), 3, 4))
+  unique_dates <- tibble::tibble(dates = unique(date)) %>%
+
+  dplyr::mutate(fin_year = paste0(ifelse(lubridate::month(dates) >= 4,
+                                  lubridate::year(dates),
+                                  lubridate::year(dates) - 1), "/",
+                           substr(ifelse(lubridate::month(dates) >= 4,
+                                         lubridate::year(dates) + 1,
+                                         lubridate::year(dates)), 3, 4))) %>%
+    dplyr::right_join(tibble::tibble(dates = date))
+
+  unique_dates$fin_year
+
```

Comment:
I would write the function as:

```r
fin_year <- function(date) {

  if (!inherits(date, c("Date", "POSIXct"))) {
    stop("The input must have Date or POSIXct class.")
  }

  tibble::tibble(dates = unique(date)) %>%
    dplyr::mutate(fin_year = paste0(ifelse(lubridate::month(.data$dates) >= 4,
                                           lubridate::year(.data$dates),
                                           lubridate::year(.data$dates) - 1),
                                    "/",
                                    substr(
                                      ifelse(lubridate::month(.data$dates) >= 4,
                                             lubridate::year(.data$dates) + 1,
                                             lubridate::year(.data$dates)),
                                      3, 4))) %>%
    dplyr::right_join(tibble::tibble(dates = date), by = "dates") %>%
    dplyr::pull(fin_year)

}
```

Some of those changes are semantics (80 character line limits :eyes:). The important parts are adding `by = "dates"` to the join to prevent `'Joining, by = "dates"'` appearing in the console when the function is run, and prefixing `dates` with `.data$` to prevent the note from appearing in `R CMD check`. For that latter part to work, `rlang` needs to be imported in the `DESCRIPTION` file, and `@importFrom rlang .data` should be added to the `phsmethods.R` script, just underneath the line that imports `%>%` from `magrittr`.

More generally, I think it'd be good to give some explanation about why the code is being written in what might seem like a counter-intuitive way. It won't affect users of the function so no need to add anything to the documentation, but I think some commentary in the code itself would be useful for anyone trying to figure out what's going on.

Also, if you don't mind adding a short line to the `NEWS` file explaining that the function now runs much quicker that'd be appreciated. Technically the package is on version 0.1.1 now but I haven't done the GitHub release yet. That can wait for this and the `README` update to be merged first.